xarray/core/dataarray.py

```diff
@@ -1093,6 +1093,84 @@ def to_masked_array(self, copy=True):
         isnull = pd.isnull(self.values)
         return np.ma.MaskedArray(data=self.values, mask=isnull, copy=copy)
 
+    def to_dict(self):
+        """
+        Convert this xarray.DataArray into a dictionary following xarray
+        naming conventions.
+
+        Useful for coverting to json.
+        """
+        d = {'coords': {}, 'attrs': dict(self.attrs), 'dims': self.dims}
+
+        def time_check(val):
+            if np.issubdtype(val.dtype, np.datetime64):
+                val = val.astype('datetime64[us]')
+            elif np.issubdtype(val.dtype, np.timedelta64):
+                val = val.astype('timedelta64[us]')
+            return val
+
+        for k in self.coords:
+            data = time_check(self[k].values).tolist()
+            d['coords'].update({k: {'data': data,
+                                    'dims': self[k].dims,
+                                    'attrs': dict(self[k].attrs)}})
+
+        d.update({'data': time_check(self.values).tolist(),
+                  'name': self.name})
+        return d
+
+    @classmethod
+    def from_dict(cls, d):
+        """
+        Convert a dictionary into an xarray.DataArray
+
+        Input dict can take several forms:
+
+        d = {'dims': ('t'), 'data': x}
+
+        d = {'coords': {'t': {'dims': 't', 'data': t,
+                              'attrs': {'units':'s'}}},
+             'attrs': {'title': 'air temperature'},
+             'dims': 't',
+             'data': x,
+             'name': 'a'}
+
+        where 't' is the name of the dimesion, 'a' is the name of the array,
+        and  x and t are lists, numpy.arrays, or pandas objects.
+
+        Parameters
+        ----------
+        d : dict, with a minimum structure of {'dims': [..], 'data': [..]}
+            or optionally {'dims': [..],
+                           'data': [..],
+                           'coords': {'dims': [..], 'data': [..]}}
+
+        Returns
+        -------
+        obj : xarray.DataArray
+
+        See also
+        --------
+        xarray.Dataset.from_dict
+        """
+        coords = None
+        if 'coords' in d:
+            try:
+                coords = OrderedDict([(k, (v['dims'],
+                                           v['data'],
+                                           v.get('attrs')))
+                                      for k, v in d['coords'].items()])
+            except KeyError as e:
+                raise KeyError(
```

Comment:
The user is calling a function here to convert input rather than directly looking things up in a mapping. So we should raise a `ValueError` instead.